repair/row_level.cc

```diff
@@ -1383,14 +1393,60 @@ class repair_meta {
             // and _working_row_buf are ordered.
             utils::merge_to_gently(_working_row_buf, row_diff,
                  [this] (const repair_row& x, const repair_row& y) { return _cmp(x.boundary(), y.boundary()) < 0; });
+            for (auto& r : row_diff) {
+                thread::maybe_yield();
+                _working_row_buf_combined_hash.add(r.hash());
+            }
         }
         if (update_hash_set) {
             _peer_row_hash_sets[node_idx] = boost::copy_range<repair_hash_set>(row_diff |
                     boost::adaptors::transformed([] (repair_row& r) { thread::maybe_yield(); return r.hash(); }));
         }
-        do_apply_rows(std::move(row_diff), node_idx, update_buf).get();
+        // Repair rows in row_diff will be flushed to disk by flush_rows_in_working_row_buf,
+        // so we skip calling do_apply_rows here.
+        _flush_to_disk = true;
+    }
+    struct dk_and_mf {
+        lw_shared_ptr<const decorated_key_with_hash> dk;
+        lw_shared_ptr<mutation_fragment> mf;
+    };
+public:
+    // Must run inside a seastar thread
+    void flush_rows_in_working_row_buf() {
+        if (_flush_to_disk) {
+            _flush_to_disk = false;
+        } else {
+            return;
+        }
+        auto cmp = position_in_partition::tri_compare(*_schema);
+        lw_shared_ptr<mutation_fragment> last_mf;
+        std::list<dk_and_mf> mfs;
+        // TODO: Now repair rows from all followers are written into a
+        // single writer. Will change the multiple writer to single writer in
+        // the following patch.
+        unsigned node_idx = 0;
+        for (auto& r : _working_row_buf) {
+            thread::maybe_yield();
+            if (!r.flush_to_disk()) {
+                continue;
+            }
+            _repair_writer->create_writer(_db, node_idx);
+            auto mf = r.get_mutation_fragment_ptr();
+            if (last_mf && cmp(last_mf->position(), mf->position()) == -1 && last_mf->mergeable_with(*mf)) {
+                last_mf->apply(*_schema, std::move(*mf));
+            } else {
+                last_mf = mf;
+                mfs.push_back(dk_and_mf{r.get_dk_with_hash(), mf});
+            }
+        }
+        while (!mfs.empty()) {
+            auto& dk_mf = mfs.front();
```

Comment:
It's nicer to write the mf (it is really a dangerous acronym) in the loop above. As soon as we discover that mf is not mergeable with last_mf, we can write last_mf. This makes the writing parallel with the computation.

Reply:
Why is mf a dangerous acronym? I will take a look. I thought about it too because it saves a loop.

Reply:
https://www.youtube.com/watch?v=c0LBi1MHoaU

Reply:
It is a terrible interpretation of this lovely data structure ;-)